File: Trees/144_Binary_Tree_Preorder_Traversal/solution144.cpp
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> stack;
        TreeNode* cur = root;

        while (cur || !stack.empty()) {
            if (cur) {
                res.push_back(cur->val);
                stack.push(cur->right);
                cur = cur->left;
            } else {
                cur = stack.top();
                stack.pop();
            }
        }
        
        return res;
    }
};

// Runtime1: 0ms, beats 100%
// Memory1: 10.20MB, beats 39.38%

// Runtime2: 0ms, beats 100%
// Memory2: 10.02MB, beats 78.24%

// Runtime3: 0ms, beats 100%
// Memory3: 10.01MB, beats 78.24%

File: Trees/144_Binary_Tree_Preorder_Traversal/solution144.py
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        curr, stack = root, []
        res = []
        while curr or stack:
            if curr:
                res.append(curr.val)
                stack.append(curr.right)
                curr = curr.left
            else:
                curr = stack.pop()
        return res

# Runtime1: 32ms, beats 75.35%
# Memory1: 16.48MB, beats 71.07%

# Runtime2: 26ms, beats 96.30%
# Memory2: 16.56MB, beats 22.17%

# Runtime3: 20ms, beats 99.66%
# Memory3: 16.57MB, beats 22.17%